deducers.go

```diff
@@ -0,0 +1,292 @@
+	select {
+	case sa := <-retchans.ret:
+		return sa, nil
+	case err := <-retchans.err:
+		return nil, err
+	}
+}
+
+type sourceCoordinator struct {
+	callMgr   *callManager
+	srcmut    sync.RWMutex // guards srcs and nameToURL maps
+	srcs      map[string]*sourceGateway
+	nameToURL map[string]string
+	psrcmut   sync.Mutex // guards protoSrcs map
+	protoSrcs map[string][]srcReturnChans
+	deducer   *deductionCoordinator
+	cachedir  string
+}
+
+func newSourceCoordinator(cm *callManager, deducer *deductionCoordinator, cachedir string) *sourceCoordinator {
+	return &sourceCoordinator{
+		callMgr:   cm,
+		deducer:   deducer,
+		cachedir:  cachedir,
+		srcs:      make(map[string]*sourceGateway),
+		nameToURL: make(map[string]string),
+		protoSrcs: make(map[string][]srcReturnChans),
+	}
+}
+
+func (sc *sourceCoordinator) getSourceGatewayFor(ctx context.Context, id ProjectIdentifier) (*sourceGateway, error) {
+	normalizedName := id.normalizedSource()
+
+	sc.srcmut.RLock()
+	if url, has := sc.nameToURL[normalizedName]; has {
+		if srcGate, has := sc.srcs[url]; has {
+			sc.srcmut.RUnlock()
+			return srcGate, nil
+		}
+	}
+	sc.srcmut.RUnlock()
+
+	// No gateway exists for this path yet; set up a proto, being careful to fold
+	// together simultaneous attempts on the same path.
+	rc := srcReturnChans{
+		ret: make(chan *sourceGateway),
+		err: make(chan error),
+	}
+
+	// The rest of the work needs its own goroutine, the results of which will
+	// be re-joined to this call via the return chans.
+	go func() {
+		sc.psrcmut.Lock()
+		if chans, has := sc.protoSrcs[normalizedName]; has {
+			// Another goroutine is already working on this normalizedName. Fold
+			// in with that work by attaching our return channels to the list.
+			sc.protoSrcs[normalizedName] = append(chans, rc)
+			sc.psrcmut.Unlock()
+			return
+		}
+
+		sc.protoSrcs[normalizedName] = []srcReturnChans{rc}
+		sc.psrcmut.Unlock()
+
+		doReturn := func(sa *sourceGateway, err error) {
+			sc.psrcmut.Lock()
+			if sa != nil {
+				for _, rc := range sc.protoSrcs[normalizedName] {
+					rc.ret <- sa
+				}
+			} else if err != nil {
+				for _, rc := range sc.protoSrcs[normalizedName] {
+					rc.err <- err
+				}
+			} else {
+				panic("sa and err both nil")
+			}
+
+			delete(sc.protoSrcs, normalizedName)
+			sc.psrcmut.Unlock()
+		}
+
+		pd, err := sc.deducer.deduceRootPath(normalizedName)
+		if err != nil {
+			// As in the deducer, don't cache errors so that externally-driven retry
+			// strategies can be constructed.
+			doReturn(nil, err)
+			return
+		}
+
+		// It'd be quite the feat - but not impossible - for an actor
+		// corresponding to this normalizedName to have slid into the main
+		// sources map after the initial unlock, but before this goroutine got
+		// scheduled. Guard against that by checking the main sources map again
+		// and bailing out if we find an entry.
+		sc.srcmut.RLock()
+		srcGate, has := sc.srcs[normalizedName]
+		sc.srcmut.RUnlock()
+		if has {
```

Comment:
>  I don't see how it would be possible for `deduceRootPath` to be run multiple times for the same path.

Suppose `getSourceGatewayFor` is called twice with the same arguments in two different goroutines, G1 and G2.

G1 checks whether does the L164 logic - it grabs the srcmut read lock, sees that sc.nameToURL doesn't have normalizedName, so it releases the lock. Now G2 is scheduled, and does exactly the same thing.

G1 starts this goroutine - the one from L182 to L272. It grabs the psrcmut write lock, sees that there is no entry in `p.protoSrcs`, so it creates one. It continues doing work, including running `deduceRootPath`, all the way to running `doReturn`. Again, it grabs the write lock, sends its output on the right channel, and then calls `delete(sc.protoSrcs, normalizedName)`. It releases the lock and exits.

Now G2 is scheduled. It starts at the top of the big goroutine. It grabs the psrcmut lock. It checks if `sc.protoSrcs` has normalized name. It doesn't, because G1 deleted its entry. So, G2 continues - it will run deduceRootPath again.

So, there we are. Certainly not a very common scheduling, but definitely possible.